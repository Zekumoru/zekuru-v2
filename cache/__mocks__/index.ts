import { sampleChannelLinks, sampleTranslateChannels } from './sample-data';

export = {
  translateChannel: {
    get: jest
      .fn()
      .mockImplementation(async (channelId) =>
        sampleTranslateChannels.find((trChannel) => trChannel.id === channelId)
      ),
    set: jest.fn(),
  },
  channelLink: {
    create: jest
      .fn()
      .mockImplementation(async (channelId) =>
        sampleChannelLinks.find((chLink) => chLink.id === channelId)
      ),
    get: jest
      .fn()
      .mockImplementation(async (channelId) =>
        sampleChannelLinks.find((chLink) => chLink.id === channelId)
      ),
    update: jest.fn(),
  },
};
