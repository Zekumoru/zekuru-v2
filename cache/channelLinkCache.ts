import { Collection } from 'discord.js';
import ChannelLink, { IChannelLink } from '../db/models/ChannelLink';

const cacheLinks = new Collection<string, IChannelLink>();

const fetchOrCreateLink = async (channelId: string, guildId: string) => {
  // fetch from db, if not exists, create new
  const link = await ChannelLink.findOne({ id: channelId })
    .populate('links')
    .exec();
  if (link) return link;

  const newLink = new ChannelLink({
    // guildId,
    id: channelId,
    links: [],
  });
  await newLink.save();
  return newLink;
};

const create = async (channelId: string, guildId: string) => {
  // create otherwise fetch
  const link = await fetchOrCreateLink(channelId, guildId);

  cacheLinks.set(channelId, {
    _id: link._id,
    id: link.id,
    guildId: link.guildId,
    links: link.links,
    createdAt: link.createdAt,
  });
  return cacheLinks.get(channelId)!;
};

const get = async (channelId: string) => {
  // get from cache
  const linkCache = cacheLinks.get(channelId);
  if (linkCache) return linkCache;

  // if not exists, fetch from db and set to cache
  const link = await ChannelLink.findOne({ id: channelId })
    .populate('links')
    .exec();
  if (!link) return null;

  cacheLinks.set(channelId, {
    _id: link._id,
    id: link.id,
    guildId: link.guildId,
    links: link.links,
    createdAt: link.createdAt,
  });
  return cacheLinks.get(channelId)!;
};

const update = async (channelLink: IChannelLink) => {
  // update in db
  const link = await fetchOrCreateLink(channelLink.id, channelLink.guildId);
  link.overwrite(channelLink);
  await link.save();

  // update cache
  cacheLinks.set(channelLink.id, {
    _id: link._id,
    id: link.id,
    guildId: link.guildId,
    links: link.links,
    createdAt: link.createdAt,
  });
};

const deleteLink = async (channelId: string) => {
  // check if it exists first
  if ((await get(channelId)) == null) return;

  // delete from db
  await ChannelLink.deleteOne({ id: channelId });

  // remove from cache
  cacheLinks.delete(channelId);
};

const clear = () => {
  cacheLinks.clear();
};

export default { create, update, get, delete: deleteLink, clear };
