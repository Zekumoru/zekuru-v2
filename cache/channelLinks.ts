import { Collection } from 'discord.js';
import ChannelLink, { IChannelLink } from '../models/ChannelLink';

const cacheLinks = new Collection<string, IChannelLink>();

const getLink = async (channelId: string, guildId: string) => {
  // fetch from db, if not exists, create new
  const link = await ChannelLink.findOne({ id: channelId })
    .populate('links')
    .exec();
  if (link) return link;

  const newLink = new ChannelLink({
    guildId,
    id: channelId,
    links: [],
  });
  await newLink.save();
  return newLink;
};

const create = async (channelId: string, guildId: string) => {
  // create otherwise fetch
  const link = await getLink(channelId, guildId);

  cacheLinks.set(channelId, {
    guildId,
    _id: link._id,
    id: link.id,
    links: link.links,
    createdAt: link.createdAt,
  });
  return cacheLinks.get(channelId)!;
};

const update = async (channelLink: IChannelLink) => {
  // update in db
  const link = await getLink(channelLink.id, channelLink.guildId);
  link.overwrite(channelLink);
  await link.save();

  // update cache
  cacheLinks.set(
    channelLink.id,
    await getLink(channelLink.id, channelLink.guildId)
  );
};

const get = async (channelId: string) => {
  // get from cache
  const linkCache = cacheLinks.get(channelId);
  if (linkCache) return linkCache;

  // if not exists, fetch from db and set to cache
  const link = await ChannelLink.findOne({ id: channelId })
    .populate('links')
    .exec();
  if (!link) return null;

  cacheLinks.set(channelId, {
    _id: link._id,
    id: link.id,
    guildId: link.guildId,
    links: link.links,
    createdAt: link.createdAt,
  });
  return cacheLinks.get(channelId)!;
};

const deleteLink = async (channelId: string) => {
  // check if it exists first
  if ((await get(channelId)) == null) return;

  // delete from db
  await ChannelLink.deleteOne({ id: channelId });

  // remove from cache
  cacheLinks.delete(channelId);
};

export default { create, update, get, delete: deleteLink };
