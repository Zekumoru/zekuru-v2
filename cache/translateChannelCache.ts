import { Collection } from 'discord.js';
import type { ITranslateChannel } from '../db/models/TranslateChannel';
import TranslateChannel from '../db/models/TranslateChannel';
import { SourceLanguageCode, TargetLanguageCode } from 'deepl-node';

const cacheTrChannels = new Collection<string, ITranslateChannel>();

const fetchOrCreateTrChannel = async (
  channelId: string,
  guildId: string,
  sourceLang: string,
  targetLang: string
) => {
  // fetch from db, if not exists, create new
  const channel = await TranslateChannel.findOne({ id: channelId });
  if (channel) return channel;

  const newChannel = new TranslateChannel({
    id: channelId,
    guildId,
    sourceLang,
    targetLang,
  });
  await newChannel.save();
  return newChannel;
};

const set = async (
  channelId: string,
  guildId: string,
  sourceLang: string,
  targetLang: string
) => {
  // update in db
  const channel = await fetchOrCreateTrChannel(
    channelId,
    guildId,
    sourceLang,
    targetLang
  );
  channel.overwrite({
    id: channelId,
    guildId: channel.guildId,
    sourceLang,
    targetLang,
    createdAt: channel.createdAt,
  });
  await channel.save();

  // update cache
  cacheTrChannels.set(channelId, {
    _id: channel._id,
    id: channel.id,
    guildId: channel.guildId,
    targetLang: channel.targetLang as TargetLanguageCode,
    sourceLang: channel.sourceLang as SourceLanguageCode,
    createdAt: channel.createdAt,
  });
};

const get = async (channelId: string) => {
  // get from cache
  const channelCache = cacheTrChannels.get(channelId);
  if (channelCache) return channelCache;

  // if not exists, fetch from db and set to cache
  const channel = await TranslateChannel.findOne({ id: channelId });
  if (!channel) return null;

  // update cache
  cacheTrChannels.set(channelId, {
    _id: channel._id,
    id: channel.id,
    guildId: channel.guildId,
    targetLang: channel.targetLang as TargetLanguageCode,
    sourceLang: channel.sourceLang as SourceLanguageCode,
    createdAt: channel.createdAt,
  });
  return cacheTrChannels.get(channelId)!;
};

const unset = async (channelId: string) => {
  // check if it exists first
  if ((await get(channelId)) == null) return;

  // delete from db
  await TranslateChannel.deleteOne({ id: channelId });

  // remove from cache
  cacheTrChannels.delete(channelId);
};

export default { set, unset, get };
