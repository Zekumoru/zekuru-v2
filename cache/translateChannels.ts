import { Collection } from 'discord.js';
import type { ITranslateChannel } from '../models/TranslateChannel';
import TranslateChannel from '../models/TranslateChannel';
import { SourceLanguageCode, TargetLanguageCode } from 'deepl-node';

const channels = new Collection<string, ITranslateChannel>();

const getChannel = async (
  channelId: string,
  guildId: string,
  sourceLang: string,
  targetLang: string
) => {
  // fetch from db, if not exists, create new
  const channel = await TranslateChannel.findOne({ id: channelId });
  if (channel) return channel;

  const newChannel = new TranslateChannel({
    id: channelId,
    guildId,
    sourceLang,
    targetLang,
  });
  await newChannel.save();
  return newChannel;
};

const set = async (
  channelId: string,
  guildId: string,
  sourceLang: string,
  targetLang: string
) => {
  // update in db
  const channel = await getChannel(channelId, guildId, sourceLang, targetLang);
  channel.overwrite({
    id: channelId,
    guildId: channel.guildId,
    sourceLang,
    targetLang,
    createdAt: channel.createdAt,
  });
  await channel.save();

  // update cache
  channels.set(channelId, {
    _id: channel._id,
    id: channel.id,
    guildId: channel.guildId,
    targetLang: channel.targetLang as TargetLanguageCode,
    sourceLang: channel.sourceLang as SourceLanguageCode,
    createdAt: channel.createdAt,
  });
};

const get = async (channelId: string) => {
  // get from cache
  const channelCache = channels.get(channelId);
  if (channelCache) return channelCache;

  // if not exists, fetch from db and set to cache
  const channel = await TranslateChannel.findOne({ id: channelId });
  if (!channel) return null;

  // update cache
  channels.set(channelId, {
    _id: channel._id,
    id: channel.id,
    guildId: channel.guildId,
    targetLang: channel.targetLang as TargetLanguageCode,
    sourceLang: channel.sourceLang as SourceLanguageCode,
    createdAt: channel.createdAt,
  });
  return channels.get(channelId)!;
};

const unset = async (channelId: string) => {
  // check if it exists first
  if ((await get(channelId)) == null) return;

  // delete from db
  await TranslateChannel.deleteOne({ id: channelId });

  // remove from cache
  channels.delete(channelId);
};

export default { set, unset, get };
