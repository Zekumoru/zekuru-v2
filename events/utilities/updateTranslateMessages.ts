import { ChannelType, Message, PartialMessage } from 'discord.js';
import MessageLink from '../../db/models/MessageLink';
import cache from '../../cache';
import getMessagesFromMessageLink from '../../commands/utilities/getMessagesFromMessageLink';
import { errorDebug } from '../../utilities/logger';
import { DISCORD_MESSAGE_CHARS_LIMIT } from '../translation/limits';
import { codeLanguagesMap } from '../../translation/languages';
import buildEmojisOnlyMap from '../translation/buildEmojisOnlyMap';
import { ITranslateChannel } from '../../db/models/TranslateChannel';
import getLanguagesFromTrChannels from '../translation/getLanguagesFromTrChannels';
import tagTranscoder from '../../utilities/tagTranscoder';

const updateTranslateMessages = async (
  newMessage: Message<boolean> | PartialMessage
) => {
  if (!newMessage.guild) return;
  if (!newMessage.content) return;

  // get message link
  const link = await MessageLink.findOne({ messageId: newMessage.id });
  if (!link) return;

  // check if the edited message's translate channel exists
  const sourceTrChannel = await cache.translateChannel.get(
    newMessage.channelId
  );
  if (!sourceTrChannel) return;

  const translator = await cache.translator.get(newMessage.guild.id);
  if (!translator) return;

  const sourceLanguage = codeLanguagesMap.get(sourceTrChannel.languageCode);
  if (!sourceLanguage) return;

  // get other channels' linked messages
  const messages = (
    await getMessagesFromMessageLink(link, newMessage.id, newMessage.guild)
  )?.filter((message) => {
    // make sure it's sent by the webhook using webhookId
    return (
      message &&
      message.webhookId &&
      message.channel.type === ChannelType.GuildText
    );
  }) as Message[] | undefined;
  if (!messages) return;

  const targetTrChannels = (
    await Promise.all(
      messages.map((message) => {
        return cache.translateChannel.get(message.channelId);
      })
    )
  ).filter(Boolean) as ITranslateChannel[];

  const emojisOnlyMap = buildEmojisOnlyMap(newMessage, targetTrChannels);

  // build warnings strings
  const warningsStrBuilder: string[] = [];
  let results: { [key: string]: string | undefined } = {};

  // translate if not emoji-only
  if (!emojisOnlyMap) {
    // because Discord has 2K characters limit, trim it to 2K
    const userReachedLimit =
      newMessage.content.length > DISCORD_MESSAGE_CHARS_LIMIT;

    if (userReachedLimit) {
      warningsStrBuilder.push(
        `**Warning:** You edited a message over 2000 characters. Due to Discord's characters limit, only the first 2000 characters will be translated.`
      );
    }

    const trimmedContent = userReachedLimit
      ? newMessage.content.slice(0, DISCORD_MESSAGE_CHARS_LIMIT)
      : newMessage.content;

    const [contentToTranslate, tagTable] = tagTranscoder.encode(trimmedContent);

    results = await translator.translate({
      content: contentToTranslate,
      contentLanguage: sourceLanguage,
      targetLanguages: getLanguagesFromTrChannels(targetTrChannels),
      translatorType: 'deepl',
    });

    Object.keys(results).forEach((key) => {
      const translated = results[key];
      if (!translated) return;
      // REMEMBER: Trim the translated output later since it might have
      //           surpassed Discord's 2K characters limit
      results[key] = tagTranscoder.decode(translated, tagTable);
    });
  }

  await Promise.all(
    messages.map(async (message) => {
      try {
        const trChannel = await cache.translateChannel.get(message.channelId);
        if (!trChannel) return;

        const language = codeLanguagesMap.get(trChannel.languageCode);
        if (!language) return;

        const channel = newMessage.client.channels.cache.get(trChannel.id);
        if (!channel) return;
        if (channel.type !== ChannelType.GuildText) return;

        const webhook = await cache.webhook.get(channel);

        // edit old message
        let translatedContent: string | undefined;
        if (emojisOnlyMap) {
          // content is an emoji
          translatedContent = emojisOnlyMap.get(trChannel.id);
        } else {
          const result = results[language.name];
          const botReachedLimit =
            !!result && result.length > DISCORD_MESSAGE_CHARS_LIMIT;

          translatedContent = botReachedLimit
            ? result.slice(0, DISCORD_MESSAGE_CHARS_LIMIT)
            : result;

          // notify user if the bot's translated message is over 2K
          if (botReachedLimit) {
            warningsStrBuilder.push(
              `**Warning:** The translated edited message in <#${trChannel.id}> is over 2000 characters. Due to Discord's characters limit, only the first 2000 characters will be shown.`
            );
          }
        }

        await webhook.editMessage(message.id, {
          content: translatedContent,
        });

        // Send warnings
        if (warningsStrBuilder.length) {
          await newMessage.reply({
            content: warningsStrBuilder.join('\n'),
          });
        }
      } catch (error) {
        errorDebug(error);
      }
    })
  );
};

export default updateTranslateMessages;
