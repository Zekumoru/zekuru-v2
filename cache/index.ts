import channelLinkCache from './channelLinkCache';
import translateChannelCache from './translateChannelCache';
import translatorCache from './translatorCache';
import webhookCache from './webhookCache';

export default {
  channelLink: channelLinkCache,
  translateChannel: translateChannelCache,
  translator: translatorCache,
  webhook: webhookCache,
};
