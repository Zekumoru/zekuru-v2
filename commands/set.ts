import {
  ActionRowBuilder,
  AutocompleteInteraction,
  ButtonBuilder,
  ButtonStyle,
  ChatInputCommandInteraction,
  PermissionFlagsBits,
  SlashCommandBuilder,
} from 'discord.js';
import { createCommand } from '../types/DiscordCommand';
import { sourceLanguages, targetLanguages } from '../translation/languages';
import translateChannels from '../cache/translateChannels';
import { updateTranslateMessages } from '../events/messageUpdateTranslate';

const data = new SlashCommandBuilder()
  .setName('set')
  .setDescription(`Set a channel's language.`)
  .addStringOption((option) =>
    option
      .setName('language')
      .setDescription('The language to set.')
      .setAutocomplete(true)
      .setRequired(true)
  )
  .addChannelOption((option) =>
    option
      .setName('channel')
      .setDescription('The channel to set the language of.')
      .setRequired(false)
  )
  .setDefaultMemberPermissions(PermissionFlagsBits.ManageGuild);

const autocomplete = async (interaction: AutocompleteInteraction) => {
  const focusedValue = interaction.options.getFocused();
  const choices = sourceLanguages.map((lang) => lang.name);
  const filtered = choices
    .filter((choice) =>
      choice.toLowerCase().startsWith(focusedValue.toLowerCase())
    )
    .slice(0, 25);

  await interaction.respond(
    filtered.map((choice) => ({ name: choice, value: choice }))
  );
};

const execute = async (interaction: ChatInputCommandInteraction) => {
  if (!interaction.guildId) {
    await interaction.reply({
      content: `This command is only available on servers.`,
    });
    return;
  }

  const guildId = interaction.guildId;
  const channelId =
    interaction.options.getChannel('channel')?.id ?? interaction.channelId;

  const language = interaction.options.getString('language');
  if (!language) {
    await interaction.reply({
      content: `Please specify a language.`,
    });
    return;
  }

  const sourceLang = sourceLanguages.find((lang) =>
    lang.name.includes(language)
  )?.code;
  if (!sourceLang) {
    await interaction.reply({
      content: `Invalid language '${language}'.`,
    });
    return;
  }

  const targetLang = targetLanguages.find((lang) =>
    lang.name.includes(language)
  )?.code;
  if (!targetLang) {
    await interaction.reply({
      content: `Error! Target language is missing. Please contact the developer.`,
    });
    return;
  }

  const trChannel = await translateChannels.get(channelId);

  if (trChannel) {
    const oldLanguage =
      sourceLanguages.find((lang) => lang.code === trChannel.sourceLang)
        ?.name ?? trChannel.sourceLang;

    // if already set with the given language
    if (oldLanguage === language) {
      await interaction.reply({
        content: `<#${channelId}> is already set to \`${language}\`!`,
      });
      return;
    }

    // start asking if change the channel's language
    const confirm = new ButtonBuilder()
      .setCustomId('confirm')
      .setLabel('Yes, change')
      .setStyle(ButtonStyle.Danger);

    const cancel = new ButtonBuilder()
      .setCustomId('cancel')
      .setLabel('Cancel')
      .setStyle(ButtonStyle.Secondary);

    const row = new ActionRowBuilder<ButtonBuilder>().addComponents(
      cancel,
      confirm
    );

    const response = await interaction.reply({
      content: `<#${channelId}> is already set to \`${oldLanguage}\`. Do you want to change it to \`${language}\`?`,
      components: [row],
    });

    try {
      const confirmation = await response.awaitMessageComponent({
        filter: (i) => i.user.id === interaction.user.id,
        time: 30_000,
      });

      if (confirmation.customId === 'confirm') {
        await translateChannels.set(channelId, guildId, sourceLang, targetLang);
        await confirmation.update({
          content: `<#${channelId}> has been changed to \`${language}\`.`,
          components: [],
        });
      } else if (confirmation.customId === 'cancel') {
        await confirmation.update({
          content: `<#${channelId}>'s language not changed`,
          components: [],
        });
      }

      // reflect updated message to translation channels
      await updateTranslateMessages(confirmation.message);
    } catch (error) {
      const message = await interaction.editReply({
        content: `Confirmation not received within 30 seconds, cancelling.`,
        components: [],
      });

      // reflect updated message to translation channels
      await updateTranslateMessages(message);
    }

    return;
  }

  // set channel's language for the first time
  await translateChannels.set(channelId, guildId, sourceLang, targetLang);
  await interaction.reply({
    content: `<#${channelId}> has been set to \`${language}\`.`,
  });
};

export default createCommand({
  data,
  autocomplete,
  execute,
});
