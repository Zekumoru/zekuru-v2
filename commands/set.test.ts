import set from './set';
import createMockInteraction from '../test/createMockChatInputCommandInteraction';
import cache from '../cache';
import updateTranslateMessages from '../events/utilities/updateTranslateMessages';
import createMockInteractionResponse from '../test/createMockInteractionResponse';

jest.mock('../translation/languages.ts');
jest.mock('../cache');
jest.mock('../events/utilities/updateTranslateMessages');

describe('/set command', () => {
  const mockCache = jest.mocked(cache);
  const mockUpdateTranslateMessages = jest.mocked(updateTranslateMessages);

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should notify user that the command has been called outside of a server', async () => {
    const interaction = createMockInteraction({ guildId: undefined });

    await set.execute(interaction);

    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
  });

  it(`should notify user if they didn't provide the language to set the channel with`, async () => {
    const interaction = createMockInteraction();
    interaction.options.getString.mockReturnValue(null);

    await set.execute(interaction);

    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
  });

  it(`should notify user if they didn't provide correct language name`, async () => {
    const interaction = createMockInteraction();
    interaction.options.getString.mockReturnValue('gibberish');

    await set.execute(interaction);

    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
  });

  it(`should notify user if the provided language is missing its target language (DeepL)`, async () => {
    const interaction = createMockInteraction();
    interaction.options.getString.mockReturnValue('English (GB)');

    await set.execute(interaction);

    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
  });

  it(`should set language`, async () => {
    const interaction = createMockInteraction();
    interaction.options.getChannel.mockReturnValue({ id: '000' } as any);
    interaction.options.getString.mockReturnValue('English');

    await set.execute(interaction);

    expect(mockCache.translateChannel.set).toHaveBeenCalled();
    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
  });

  it(`should change language`, async () => {
    const interaction = createMockInteraction();
    interaction.options.getChannel.mockReturnValue({ id: '100' } as any);
    interaction.options.getString.mockReturnValue('Japanese');
    const mockUpdate = jest.fn();
    interaction.reply.mockResolvedValue(
      createMockInteractionResponse(
        {},
        { customId: 'confirm', update: mockUpdate }
      )
    );

    await set.execute(interaction);

    expect(mockCache.translateChannel.set).toHaveBeenCalled();
    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
    expect(mockUpdate.mock.calls[0][0]).toMatchSnapshot();
  });

  it(`should not change language if user cancelled`, async () => {
    const interaction = createMockInteraction();
    interaction.options.getChannel.mockReturnValue({ id: '100' } as any);
    interaction.options.getString.mockReturnValue('Japanese');
    const mockUpdate = jest.fn();
    interaction.reply.mockResolvedValue(
      createMockInteractionResponse(
        {},
        { customId: 'cancel', update: mockUpdate }
      )
    );

    await set.execute(interaction);

    expect(mockCache.translateChannel.set).not.toHaveBeenCalled();
    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
    expect(mockUpdate.mock.calls[0][0]).toMatchSnapshot();
  });

  it(`should not change language if timeout reached`, async () => {
    const interaction = createMockInteraction();
    interaction.options.getChannel.mockReturnValue({ id: '100' } as any);
    interaction.options.getString.mockReturnValue('Japanese');
    interaction.reply.mockResolvedValue({
      awaitMessageComponent: jest.fn().mockImplementation(async () => {
        throw new Error('generic error');
      }),
    } as any);

    await set.execute(interaction);

    expect(mockCache.translateChannel.set).not.toHaveBeenCalled();
    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
    expect(interaction.editReply.mock.calls[0][0]).toMatchSnapshot();
  });
});
