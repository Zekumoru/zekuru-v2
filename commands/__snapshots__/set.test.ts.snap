// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/set command should change language 1`] = `
{
  "components": [
    {
      "components": [
        {
          "custom_id": "cancel",
          "emoji": undefined,
          "label": "Cancel",
          "style": 2,
          "type": 2,
        },
        {
          "custom_id": "confirm",
          "emoji": undefined,
          "label": "Yes, change",
          "style": 4,
          "type": 2,
        },
      ],
      "type": 1,
    },
  ],
  "content": "<#100> is already set to \`English\`. Do you want to change it to \`Japanese\`?",
}
`;

exports[`/set command should change language 2`] = `
{
  "components": [],
  "content": "<#100> has been changed to \`Japanese\`.",
}
`;

exports[`/set command should not change language if timeout reached 1`] = `
{
  "components": [
    {
      "components": [
        {
          "custom_id": "cancel",
          "emoji": undefined,
          "label": "Cancel",
          "style": 2,
          "type": 2,
        },
        {
          "custom_id": "confirm",
          "emoji": undefined,
          "label": "Yes, change",
          "style": 4,
          "type": 2,
        },
      ],
      "type": 1,
    },
  ],
  "content": "<#100> is already set to \`English\`. Do you want to change it to \`Japanese\`?",
}
`;

exports[`/set command should not change language if timeout reached 2`] = `
{
  "components": [],
  "content": "Confirmation not received within 30 seconds, cancelling.",
}
`;

exports[`/set command should not change language if user cancelled 1`] = `
{
  "components": [
    {
      "components": [
        {
          "custom_id": "cancel",
          "emoji": undefined,
          "label": "Cancel",
          "style": 2,
          "type": 2,
        },
        {
          "custom_id": "confirm",
          "emoji": undefined,
          "label": "Yes, change",
          "style": 4,
          "type": 2,
        },
      ],
      "type": 1,
    },
  ],
  "content": "<#100> is already set to \`English\`. Do you want to change it to \`Japanese\`?",
}
`;

exports[`/set command should not change language if user cancelled 2`] = `
{
  "components": [],
  "content": "<#100>'s language not changed",
}
`;

exports[`/set command should notify user if the provided language is missing its target language (DeepL) 1`] = `
{
  "content": "Error! Target language for 'English (GB)' is missing. Please contact the developer.",
}
`;

exports[`/set command should notify user if they didn't provide correct language name 1`] = `
{
  "content": "Invalid language 'gibberish'.",
}
`;

exports[`/set command should notify user if they didn't provide the language to set the channel with 1`] = `
{
  "content": "Please specify a language.",
}
`;

exports[`/set command should notify user that the command has been called outside of a server 1`] = `
{
  "content": "This command is only available on servers.",
}
`;

exports[`/set command should set language 1`] = `
{
  "content": "<#000> has been set to \`English\`.",
}
`;
